### PR TITLE
CSCFAIRMETA-639-Issued-date

### PR DESCRIPTION
**Pull request by katrite**
Make issued_date mandatory field for ida and att-catalogs
If issued_date is not set, todays date is filled in
If issued_date is empty, error is returned

### CODE REVIEW

File: src/metax_api/tests/api/rest/base/views/datasets/write.py [hunk 1]
```diff
@@ -195,7 +195,7 @@ def _set_cr_owner_and_state(self, cr_id, state, owner):
         cr.data_catalog_id = DataCatalog.objects.get(catalog_json__identifier=END_USER_ALLOWED_DATA_CATALOGS[0]).id
         cr.force_save()
 
-    def test_field_exists(self):
+    def test_field_state_exists(self):
```
Comment (katrite) [inline src/metax_api/tests/api/rest/base/views/datasets/write.py:198]: This is not directly related, but I realized the old test name is not telling too much, especially now when we are testing that another field ('issued') exists

File: src/metax_api/models/catalog_record.py [hunk 2]
```diff
@@ -1195,6 +1199,13 @@ def _pre_create_operations(self, pid_type=None):
 
             self.date_cumulation_started = self.date_created
 
+        if (self.catalog_is_ida() or self.catalog_is_att()):
```
Comment (tompulli) [inline src/metax_api/models/catalog_record.py:1202]: If I understood correctly, the issued should be auto-filled for every dataset except for harvested?

File: src/metax_api/models/catalog_record.py [hunk 3]
```diff
@@ -1195,6 +1199,13 @@ def _pre_create_operations(self, pid_type=None):
...
+            if 'issued' not in self.research_dataset:
+                try:
```
Comment (tompulli) [inline src/metax_api/models/catalog_record.py:1204]: I don't think try-except is necessary here since we are only using internal values and functions

File: src/metax_api/models/catalog_record_v2.py [hunk 4]
```diff
@@ -163,6 +165,14 @@ def _save_as_draft(self):
         from metax_api.services import CommonService
         return CommonService.get_boolean_query_param(self.request, 'draft')
 
+    def _handle_issued_date(self):
```
Comment (tompulli) [inline src/metax_api/models/catalog_record_v2.py:168]: this function can be defined in V1 model so it can be utilized by both models

File: src/metax_api/models/catalog_record_v2.py [hunk 5]
```diff
@@ -126,6 +126,8 @@ def _post_create_operations(self, pid_type=None):
             else:
                 self.use_doi_for_published = False
 
+            self._handle_issued_date()
```
Comment (tompulli) [inline src/metax_api/models/catalog_record_v2.py:129]: This should be done in pre_create_operations for all datasets. Now it is actually not saved for datasets that are published straight away.

File: src/metax_api/tests/api/rest/v2/views/datasets/write.py [hunk 6]
```diff
@@ -180,6 +181,24 @@ class CatalogRecordApiWriteCreateTests(CatalogRecordApiWriteCommon):
...
+        for catalog in [ATT_CATALOG, IDA_CATALOG]:
+            dc = DataCatalog.objects.get(pk=2)
+            dc.catalog_json['identifier'] = catalog
+            dc.force_save()
+
+            # Create a catalog record in att catalog
+            self.cr_test_data['data_catalog'] = dc.catalog_json
+            self.cr_test_data['research_dataset'].pop('issued', None)
+
+            response = self.client.post('/rest/datasets', self.cr_test_data, format="json")
```
Comment (tompulli) [inline src/metax_api/tests/api/rest/v2/views/datasets/write.py:199]: v1 endpoint

File: src/metax_api/models/catalog_record.py (hunk 2 above)
Comment (katrite) [reply src/metax_api/models/catalog_record.py:1202]: Good point, now i modified it to be mandatory for all except harvested and drafts...

File: src/metax_api/models/catalog_record_v2.py (hunk 5 above)
Comment (katrite) [reply src/metax_api/models/catalog_record_v2.py:129]: Usch, you're right, same mistake again.. :E

File: src/metax_api/tests/api/rest/base/views/datasets/write.py [hunk 7]
```diff
@@ -440,6 +440,22 @@ class CatalogRecordApiWriteCreateTests(CatalogRecordApiWriteCommon):
     #
     #
 
+    def test_issued_date_is_mandatory(self):
```
Comment (tompulli) [inline src/metax_api/tests/api/rest/base/views/datasets/write.py:443]: 'mandatory' might be a bit misleading since the field is merely filled by metax if not provided. maybe 'issued_date_is_generated' or similar?

File: src/metax_api/models/catalog_record_v2.py [hunk 8]
```diff
@@ -99,6 +99,7 @@ def _pre_create_operations(self):
 
         self.research_dataset['metadata_version_identifier'] = generate_uuid_identifier()
         self.identifier = generate_uuid_identifier()
+        self._handle_issued_date()
```
Comment (tompulli) [inline src/metax_api/models/catalog_record_v2.py:102]: This works for datasets that are published right away, but not on datasets that are first drafts and then published. There's also third way of publishing datasets with merge_draft. It should be investigated how that works as well and it would be good to write appropriate tests for these scenarios as well. Note that these only apply for v2 API

File: src/metax_api/tests/api/rest/base/views/datasets/write.py (hunk 7 above)
Comment (katrite) [reply src/metax_api/tests/api/rest/base/views/datasets/write.py:443]: You're right, its now changed to 'generated'.

File: src/metax_api/models/catalog_record_v2.py (hunk 8 above)
Comment (katrite) [reply src/metax_api/models/catalog_record_v2.py:102]: Related to this https://jira.eduuni.fi/browse/CSCFAIRMETA-642, issued date is not generated for drafts and it is tested that they won't have this field.

File: src/metax_api/models/catalog_record.py [hunk 9]
```diff
@@ -44,6 +44,7 @@
 LEGACY_CATALOGS = settings.LEGACY_CATALOGS
 IDA_CATALOG = settings.IDA_DATA_CATALOG_IDENTIFIER
 PAS_CATALOG = settings.PAS_DATA_CATALOG_IDENTIFIER
+ATT_CATALOG = settings.ATT_DATA_CATALOG_IDENTIFIER
```
Comment (tompulli) [inline src/metax_api/models/catalog_record.py:47]: This and the catalog_is_att function is not used which causes the test coverage to fall. should be removed.

File: src/metax_api/tests/api/rest/base/views/datasets/write.py [hunk 10]
```diff
@@ -440,6 +440,22 @@ class CatalogRecordApiWriteCreateTests(CatalogRecordApiWriteCommon):
     #
     #
 
+    def test_issued_date_is_generated(self):
+        ''' Issued date is generated for all but harvested catalogs if it doesn't exists '''
+        # ATT
```
Comment (tompulli) [inline src/metax_api/tests/api/rest/base/views/datasets/write.py:445]: should be removed since we are testing also ida catalog?

File: src/metax_api/tests/api/rest/base/views/datasets/write.py [hunk 11]
```diff
@@ -440,6 +440,22 @@ class CatalogRecordApiWriteCreateTests(CatalogRecordApiWriteCommon):
...
+        for catalog in [ATT_CATALOG, IDA_CATALOG]:
+            dc = DataCatalog.objects.get(pk=2)
+            dc.catalog_json['identifier'] = catalog
+            dc.force_save()
+
+            # Create a catalog record in att catalog
```
Comment (tompulli) [inline src/metax_api/tests/api/rest/base/views/datasets/write.py:451]: nitpick: using also ida catalog

File: src/metax_api/tests/api/rest/v2/views/datasets/drafts.py [hunk 12]
```diff
@@ -64,12 +64,32 @@ def _set_cr_owner_and_state(self, cr_id, state, owner):
...
+    def test_field_state_exists(self):
         """Try fetching any dataset, field 'state' should be returned'"""
 
         cr = self.client.get('/rest/v2/datasets/13').data
         self.assertEqual('state' in cr, True)
 
+    def _test_issued_date_is_not_generated_for_drafts(self):
+        '''  Drafts will not have the issued date generated
+             Field is created when dataset is published
+        '''
```
Comment (tompulli) [inline src/metax_api/tests/api/rest/v2/views/datasets/drafts.py:76]: nitpick: multiline comments either in a inline manner (example on lines 244-245) or in a block type manner (example on lines 26-28)

File: src/metax_api/models/catalog_record.py (hunk 9 above)
Comment (katrite) [reply src/metax_api/models/catalog_record.py:47]: ....Removed

File: src/metax_api/tests/api/rest/base/views/datasets/write.py (hunk 10 above)
Comment (katrite) [reply src/metax_api/tests/api/rest/base/views/datasets/write.py:445]: Good point

File: src/metax_api/tests/api/rest/base/views/datasets/write.py (hunk 11 above)
Comment (katrite) [reply src/metax_api/tests/api/rest/base/views/datasets/write.py:451]: Good point

**Review by tompulli — APPROVED** [review]
